pkg/api/v1beta3/types.go

```diff
@@ -934,7 +934,7 @@ type ServiceSpec struct {
 	Ports []ServicePort `json:"ports" description:"ports exposed by the service"`
 
 	// This service will route traffic to pods having labels matching this selector. If null, no endpoints will be automatically created. If empty, all pods will be selected.
-	Selector map[string]string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
+	Selector string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
```

Comment:
How about an array of label requirements?  We can't accept both map and slice, but we can easily convert slice of structured params to requirements and still preserve nil.

On the other hand, I may have missed the discussion of the disadvantages of the current map format.  Is it necessary to allow more effective representation of OR?

> On Apr 20, 2015, at 9:08 PM, Brian Grant <notifications@github.com> wrote:
> 
> In pkg/api/v1beta3/types.go:
> 
> > @@ -934,7 +934,7 @@ type ServiceSpec struct {
> >  	Ports []ServicePort `json:"ports" description:"ports exposed by the service"`
> >  
> >  	// This service will route traffic to pods having labels matching this selector. If null, no endpoints will be automatically created. If empty, all pods will be selected.
> > -	Selector map[string]string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
> > +	Selector string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
> I'm pretty sure this breaks the distinction between empty and not specified. (read the description)
> 
> —
> Reply to this email directly or view it on GitHub.
>